_source/_posts/2022-06-16-oauth-java.adoc

```diff
@@ -0,0 +1,389 @@
+
+It has https://docs.spring.io/spring-security/reference/servlet/oauth2/index.html[OAuth 2.0], https://docs.spring.io/spring-security/reference/servlet/oauth2/login/index.html[OpenID Connect], and even https://docs.spring.io/spring-security/reference/servlet/saml2/index.html[SAML] support.
+
+====
+It's a shame Spring Security has to support SAML. 😞
+
+Apparently, there's a demand for it from their customers and users. Or maybe it's just mis-informed decision makers? Are developers still implementing SOAP APIs? I hope not! If not, why are developers still using SAML?!
+
+OIDC does it better, just use it!
+====
+
+== OAuth 2.0 authorization servers
+
+Up to this point, all the things I've talked about are OAuth clients. They require an identity provider to do OAuth flows. An identity provider has OAuth 2.0 authorization servers. Most of the big providers have their own proprietary implementation.
+
+However, there are a couple of open source Java-based authorization servers you can use:
+
+- https://www.keycloak.org/[Keycloak]
+- https://spring.io/projects/spring-authorization-server[Spring Authorization Server]
+- https://github.com/wso2/product-is[WSO2 Identity Server]
+
+Since I work for a company that is an OAuth 2.0 identity provider, it's my duty to inform you that using one of these is called "building your own." We recommend buying instead. From us, of course. 😉
+
+In reality, if you're a developer, you're always going to _build your own_ somewhat. Even if you use a library like Spring Security, you can't just add it to your project and expect everything to work without writing any code. Granted, it might only require 10 lines of code, but you're still building _something_.
+
+You can't just buy Okta and secure a Java API with it without writing code.
+
+== What about JavaFX?
+
+JavaFX is often touted as an excellent way to build desktop apps. The folks that recommend it are typically Java FX experts and Java enthusiasts that hate JavaScript.
+
+JavaFX's OAuth support is virtually non-existent and there doesn't seem to be many folks interested in fixing that.
+
+As one of the few Java Champions that likes JavaScript, I poked the bear a couple of weeks ago.
+
+++++
+{% twitter 1520048029641986048 %}
+++++
+
+My guess is this won't be solved anytime soon. I'd love to be wrong.
+
+// == What about Kotlin and Scala?
+// I like this ^^ title, but it seems to clash with the previous title for JavaFX
+== Do Scala and Kotlin support OAuth?
+
+Scala was all the rage in the mid-2010s. When the Play framework team announced they were writing Play 2 in Scala, the JVM world rejoiced. I was interested in it too. I took a course from Martin Odersky, learned a ton, and reached the top of Hacker News when I https://raibledesigns.com/rd/entry/play_2_0_a_web[live-blogged about it] at Devoxx Belgium in 2011.
+
+Since then, the popularity of Play has wained, Spring Boot has taken over, and Kotlin is a JVM language that lots of developers admire.
```

Comment:
```suggestion
Since then, the popularity of Play has waned, Spring Boot has taken over, and Kotlin is a JVM language that lots of developers admire.
```